editor/editor_asset_installer.cpp

```diff
@@ -82,229 +81,259 @@ void EditorAssetInstaller::open(const String &p_path, int p_depth) {
-		if (Engine::get_singleton()->has_singleton("GodotSharp")) {
-			extension_guess["cs"] = tree->get_editor_theme_icon(SNAME("CSharpScript"));
-		} else {
-			// Mark C# support as unavailable.
-			extension_guess["cs"] = tree->get_editor_theme_icon(SNAME("ImportFail"));
+			separator = source_name.find("/", separator + 1);
 		}
 
-		extension_guess["res"] = tree->get_editor_theme_icon(SNAME("Resource"));
-		extension_guess["tres"] = tree->get_editor_theme_icon(SNAME("Resource"));
-		extension_guess["atlastex"] = tree->get_editor_theme_icon(SNAME("AtlasTexture"));
-		// By default, OBJ files are imported as Mesh resources rather than PackedScenes.
-		extension_guess["obj"] = tree->get_editor_theme_icon(SNAME("Mesh"));
-
-		extension_guess["txt"] = tree->get_editor_theme_icon(SNAME("TextFile"));
-		extension_guess["md"] = tree->get_editor_theme_icon(SNAME("TextFile"));
-		extension_guess["rst"] = tree->get_editor_theme_icon(SNAME("TextFile"));
-		extension_guess["json"] = tree->get_editor_theme_icon(SNAME("TextFile"));
-		extension_guess["yml"] = tree->get_editor_theme_icon(SNAME("TextFile"));
-		extension_guess["yaml"] = tree->get_editor_theme_icon(SNAME("TextFile"));
-		extension_guess["toml"] = tree->get_editor_theme_icon(SNAME("TextFile"));
-		extension_guess["cfg"] = tree->get_editor_theme_icon(SNAME("TextFile"));
-		extension_guess["ini"] = tree->get_editor_theme_icon(SNAME("TextFile"));
-	}
+		if (!source_name.is_empty() && !asset_files.has(source_name)) {
+			asset_files.insert(source_name);
+		}
 
-	Ref<Texture2D> generic_extension = tree->get_editor_theme_icon(SNAME("Object"));
+		ret = unzGoToNextFile(pkg);
+	}
 
 	unzClose(pkg);
 
+	_check_has_toplevel();
+	// Default to false, unless forced.
+	skip_toplevel = p_autoskip_toplevel;
+	skip_toplevel_check->set_pressed(!skip_toplevel_check->is_disabled() && skip_toplevel);
+
+	_rebuild_tree();
+}
+
+void EditorAssetInstaller::_rebuild_tree() {
 	updating = true;
 	tree->clear();
+
 	TreeItem *root = tree->create_item();
 	root->set_cell_mode(0, TreeItem::CELL_MODE_CHECK);
 	root->set_checked(0, true);
-	root->set_icon(0, tree->get_theme_icon(SNAME("folder"), SNAME("FileDialog")));
+	root->set_icon(0, get_theme_icon(SNAME("folder"), SNAME("FileDialog")));
 	root->set_text(0, "res://");
 	root->set_editable(0, true);
-	HashMap<String, TreeItem *> dir_map;
 
+	HashMap<String, TreeItem *> directory_item_map;
 	int num_file_conflicts = 0;
 
-	for (const String &E : files_sorted) {
-		String path = E;
-		int depth = p_depth;
-		bool skip = false;
-		while (depth > 0) {
-			int pp = path.find("/");
-			if (pp == -1) {
-				skip = true;
-				break;
+	bool first = true;
+	for (const String &E : asset_files) {
+		if (first) {
+			first = false;
+
+			if (!toplevel_prefix.is_empty() && skip_toplevel) {
+				continue;
 			}
-			path = path.substr(pp + 1, path.length());
-			depth--;
 		}
 
-		if (skip || path.is_empty()) {
-			continue;
+		String path = E; // We're going to mutate it.
```

Comment:
You could just remove `const` from E.

Reply:
We don't want to mutate the contents of the container.

Reply:
Ah, then it's fine.
I don't think it would create copy, though not sure.